hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1439,6 +1440,238 @@ private class Emit(
+          ndPType.data.pType.stagedInitialize(aAddressDGEQRF, aNumElements.toI),
+          ndPType.copyRowMajorToColumnMajor(dataAddress, aAddressDGEQRF, M, N, mb),
+
+          tauAddress := tauPType.allocate(region, K.toI),
+          tauPType.stagedInitialize(tauAddress, K.toI),
+
+          sizeQueryAddress := ndPType.data.pType.allocate(region, 1), // One double, just to hold size query results
+
+          infoDGEQRFResult := Code.invokeScalaObject[Int, Int, Long, Int, Long, Long, Int, Int](LAPACK.getClass, "dgeqrf",
+            M.toI,
+            N.toI,
+            ndPType.data.pType.elementOffset(aAddressDGEQRF, aNumElements.toI, 0),
+            LDA.toI,
+            tauPType.elementOffset(tauAddress, K.toI, 0),
+            sizeQueryAddress,
+            -1
+          ),
+          infoDGEQRFErrorTest,
+
+          workAddress := Code.invokeStatic[Memory, Long, Long]("malloc", LWORK.toL * 8L),
+
+          infoDGEQRFResult := Code.invokeScalaObject[Int, Int, Long, Int, Long, Long, Int, Int](LAPACK.getClass, "dgeqrf",
+            M.toI,
+            N.toI,
+            ndPType.data.pType.elementOffset(aAddressDGEQRF, aNumElements.toI, 0),
+            LDA.toI,
+            tauPType.elementOffset(tauAddress, K.toI, 0),
+            workAddress,
+            LWORK
+          ),
+          Code.invokeStatic[Memory, Long, Unit]("free", workAddress.load()),
+          infoDGEQRFErrorTest
+        )
+
+        val result = if (mode == "raw") {
+          // Raw returns the column major form of A as is.
+          val rawPType = x.pType.asInstanceOf[PTuple]
+          val rawOutputSrvb = new StagedRegionValueBuilder(mb, x.pType, region)
+          val hPType = rawPType.types(0).asInstanceOf[PNDArray]
+          val tauPType = rawPType.types(1).asInstanceOf[PNDArray]
+
+          val hShapeArray = Array(N, M)
+          val hShapeBuilder = hPType.makeShapeBuilder(hShapeArray)
+          val hStridesBuilder = hPType.makeDefaultStridesBuilder(hShapeArray, mb)
+
+          val tauShapeBuilder = tauPType.makeShapeBuilder(Array(K))
+          val tauStridesBuilder = tauPType.makeDefaultStridesBuilder(Array(K), mb)
+
+          val h = hPType.construct(0, 0, hShapeBuilder, hStridesBuilder, aAddressDGEQRF, mb)
+          val tau = tauPType.construct(0, 0, tauShapeBuilder, tauStridesBuilder, tauAddress, mb)
+
+          val computeRaw = Code(
+            rawOutputSrvb.start(),
+            rawOutputSrvb.addIRIntermediate(hPType)(h),
+            rawOutputSrvb.advance(),
+            rawOutputSrvb.addIRIntermediate(tauPType)(tau),
+            rawOutputSrvb.advance(),
+            rawOutputSrvb.end()
+          )
+
+          Code(
+            alwaysNeeded,
+            computeRaw
+          )
+        }
+        else {
+          val currRow = mb.newField[Int]
+          val currCol = mb.newField[Int]
+
+          //This block assumes rDataAddress is a row major ndarray.
+          val zeroOutCorner =
```

Comment:
since rDataAddress isn't initialized until the computeR block, let's move this down to just before that block. As it stands, we have to scan past a several lines of code to see the definition of rDataAddress and the use of this val.